src/modules/update/common/update-common.c

```diff
@@ -0,0 +1,488 @@
+        return;
+    }
+
+    blob_backend = malloc(size);
+    SOL_NULL_CHECK_GOTO(blob_backend, err);
+
+    blob = sol_blob_new(SOL_BLOB_TYPE_DEFAULT, NULL, blob_backend, size);
+    SOL_NULL_CHECK_GOTO(blob, err);
+
+    memcpy(blob_backend, buf, size);
+
+    r = sol_message_digest_feed(md, blob, last);
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    sol_blob_unref(blob);
+    return;
+
+err:
+    SOL_WRN("Could not feed data to check update file hash");
+    free(blob_backend);
+    sol_blob_unref(blob);
+    sol_message_digest_del(md);
+    handle->cb_hash(handle, -EINVAL);
+}
+
+static bool
+check_hash(struct sol_update_handle *handle,
+    void (*cb)(struct sol_update_handle *handle, int status))
+{
+    struct sol_message_digest_config cfg = {
+        SOL_SET_API_VERSION(.api_version = SOL_MESSAGE_DIGEST_CONFIG_API_VERSION, )
+        .algorithm = handle->hash_algorithm,
+        .on_digest_ready = on_digest_ready_cb,
+        .on_feed_done = on_feed_done_cb,
+        .data = handle
+    };
+    struct sol_message_digest *md;
+
+    md = sol_message_digest_new(&cfg);
+    SOL_NULL_CHECK(md, false);
+
+    rewind(handle->file);
+    handle->cb_hash = cb;
+
+    /* Start feeding */
+    on_feed_done_cb(handle, md, NULL);
+
+    return true;
+}
+
+static void
+check_hash_cb(struct sol_update_handle *handle, int status)
+{
+    char *real_path;
+
+    if (status < 0) {
+        SOL_WRN("Invalid hash of update file");
+        handle->cb_fetch((void *)handle->user_data, -EINVAL, NULL);
+        goto end;
+    }
+
+    real_path = realpath(SOL_UPDATE_FILE_NAME, NULL);
+    handle->cb_fetch((void *)handle->user_data, real_path ? 0 : -ENOMEM, real_path);
+    free(real_path);
+
+end:
+    fclose(handle->file);
+    handle->file = NULL;
+    delete_handle(handle);
+}
+
+static void
+task_fetch_response(void *data, const struct sol_http_client_connection *conn,
+    struct sol_http_response *http_response)
+{
+    struct sol_update_handle *handle = data;
+    FILE *f;
+
+    handle->conn = NULL;
+
+    if (http_response->response_code != SOL_HTTP_STATUS_OK) {
+        SOL_WRN("Invalid response code from [%s] when fetching update: %d",
+            handle->url, http_response->response_code);
+        handle->cb_fetch((void *)handle->user_data, -http_response->response_code, NULL);
+        delete_handle(handle);
+        return;
+    }
+
+    if (!http_response->content.used) {
+        SOL_WRN("Empty response for fetch update");
+        handle->cb_fetch((void *)handle->user_data, -EINVAL, NULL);
+        delete_handle(handle);
+        return;
+    }
+
+    f = fopen(SOL_UPDATE_FILE_NAME, "w+e");
+    if (!f) {
+        SOL_WRN("Could not open file for writing: %s", sol_util_strerrora(errno));
+        handle->cb_fetch((void *)handle->user_data, -errno, NULL);
+        delete_handle(handle);
+        return;
+    }
+
+    fwrite(http_response->content.data, http_response->content.used, 1, f);
+    fflush(f);
```

Comment:
I'm kind of doing the `fsync` stuff, but after renaming/moving file.
And I'll update this to use streaming as soon as out http client support that.